src/meta-srv/src/selector/load_based.rs

```diff
@@ -12,234 +12,169 @@
+        let mut region_nums = Vec::with_capacity(stat_kvs.len());
+        let mut peers = Vec::with_capacity(stat_kvs.len());
+
+        for stat_k in stat_kvs.keys() {
+            let Some(stat_val) = stat_kvs.get(stat_k) else {
+                continue;
+            };
+
+            let region_num = stat_val.region_num().unwrap_or(MAX_REGION_NUMBER);
+            let Some(node_addr) = stat_val.node_addr() else {
+                continue;
+            };
+
+            let peer = Peer {
+                id: stat_k.node_id,
+                addr: node_addr,
+            };
+
+            region_nums.push(region_num);
+            peers.push(peer);
         }
 
-        let stat_keys: Vec<StatKey> = lease_kvs.keys().map(|k| k.into()).collect();
-        let stat_kvs = ctx.meta_peer_client.get_dn_stat_kvs(stat_keys).await?;
+        if region_nums.is_empty() {
+            return vec![];
+        }
 
-        let mut tuples: Vec<(LeaseKey, LeaseValue, u64)> = lease_kvs
-            .into_iter()
-            // The regions of a table need to be distributed on different datanode.
-            .filter(|(lease_k, _)| {
-                if let Some(stat_val) = stat_kvs.get(&lease_k.into()) {
-                    if let (Some(catalog), Some(schema), Some(table)) =
-                        (&ctx.catalog, &ctx.schema, &ctx.table)
-                    {
-                        return contains_table(stat_val, catalog, schema, table) != Some(true);
-                    }
-                }
-                true
-            })
-            .map(|(lease_k, lease_v)| {
-                let stat_key: StatKey = (&lease_k).into();
-
-                let region_num = match stat_kvs
-                    .get(&stat_key)
-                    .and_then(|stat_val| stat_val.region_num())
-                {
-                    Some(region_num) => region_num,
-                    None => {
-                        warn!("Failed to get stat_val by stat_key {:?}", stat_key);
-                        MAX_REGION_NUMBER
-                    }
-                };
-
-                (lease_k, lease_v, region_num)
-            })
-            .collect();
+        // unwrap safety: region_nums is not empty
+        let max_region_number = region_nums.iter().max().cloned().unwrap();
+        let min_region_number = region_nums.iter().min().cloned().unwrap();
 
-        // sort the datanodes according to the number of regions
-        tuples.sort_by(|a, b| a.2.cmp(&b.2));
+        let base_weight = match max_region_number - min_region_number {
+            0 => 1,
+            x => x,
+        };
 
-        Ok(tuples
+        peers
             .into_iter()
-            .map(|(stat_key, lease_val, _)| Peer {
-                id: stat_key.node_id,
-                addr: lease_val.node_addr,
+            .zip(region_nums)
+            .map(|(peer, region)| WeightedItem {
+                item: peer,
+                weight: (max_region_number - region + base_weight) as usize,
```

Comment:
What does `base_weight` use for? If `max_region_number` is u64::MAX, region = 0, isn't `max_region_number - region + base_weight` overflowed?

Reply:
> What does base_weight use for?

There are two uses:
1. It can prevent the weight of a certain item being 0.
2. A reasonable base_weight can prevent the gap between the maximum value and the minimum value of the weight from being too large.

> max_region_number is u64::MAX

This has some historical baggage, but it can be removed now.